File: backend/learning-service/src/main/java/com/jungle/learning/service/NoteService.java
package com.jungle.learning.service;

import com.jungle.learning.dto.NoteDTO;
import com.jungle.learning.model.Note;
import com.jungle.learning.model.ReadingProgress;
import com.jungle.learning.repository.NoteRepository;
import com.jungle.learning.repository.ReadingProgressRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class NoteService {

    private final NoteRepository noteRepository;
    private final ReadingProgressRepository progressRepository;

    @Transactional
    public NoteDTO createNote(Long progressId, Integer pageNumber, String content, 
                              String highlightedText, String color) {
        ReadingProgress progress = progressRepository.findById(progressId)
                .orElseThrow(() -> new RuntimeException("Reading progress not found"));

        Note note = new Note();
        note.setProgress(progress);
        note.setPageNumber(pageNumber);
        note.setContent(content);
        note.setHighlightedText(highlightedText);
        note.setColor(color != null ? color : "#FFEB3B");

        Note saved = noteRepository.save(note);
        return mapToDTO(saved);
    }

    @Transactional
    public NoteDTO updateNote(Long noteId, String content) {
        Note note = noteRepository.findById(noteId)
                .orElseThrow(() -> new RuntimeException("Note not found"));

        note.setContent(content);
        Note updated = noteRepository.save(note);
        return mapToDTO(updated);
    }

    @Transactional
    public void deleteNote(Long noteId) {
        noteRepository.deleteById(noteId);
    }

    public List<NoteDTO> getProgressNotes(Long progressId) {
        return noteRepository.findByProgressIdOrderByPageNumberAsc(progressId)
                .stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    public List<NoteDTO> getPageNotes(Long progressId, Integer pageNumber) {
        return noteRepository.findByProgressIdAndPageNumber(progressId, pageNumber)
                .stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    private NoteDTO mapToDTO(Note note) {
        NoteDTO dto = new NoteDTO();
        dto.setId(note.getId());
        dto.setProgressId(note.getProgress().getId());
        dto.setPageNumber(note.getPageNumber());
        dto.setContent(note.getContent());
        dto.setHighlightedText(note.getHighlightedText());
        dto.setColor(note.getColor());
        dto.setCreatedAt(note.getCreatedAt());
        dto.setUpdatedAt(note.getUpdatedAt());
        return dto;
    }
}

File: frontend/src/app/core/services/note.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Note } from '../models/ebook.model';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class NoteService {
  private apiUrl = `${environment.apiUrl}/notes`;

  constructor(private http: HttpClient) {}

  createNote(
    progressId: number,
    pageNumber: number,
    content: string,
    highlightedText?: string,
    color?: string
  ): Observable<Note> {
    return this.http.post<Note>(this.apiUrl, {
      progressId,
      pageNumber,
      content,
      highlightedText,
      color
    });
  }

  updateNote(noteId: number, content: string): Observable<Note> {
    return this.http.put<Note>(`${this.apiUrl}/${noteId}`, { content });
  }

  deleteNote(noteId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${noteId}`);
  }

  getProgressNotes(progressId: number): Observable<Note[]> {
    return this.http.get<Note[]>(`${this.apiUrl}/progress/${progressId}`);
  }

  getPageNotes(progressId: number, pageNumber: number): Observable<Note[]> {
    return this.http.get<Note[]>(`${this.apiUrl}/progress/${progressId}/page/${pageNumber}`);
  }
}
